karapace/protobuf/schema.py

```diff
@@ -106,6 +109,49 @@ def option_element_string(option: OptionElement) -> str:
     return f"option {result};\n"
 
 
+@dataclasses.dataclass(frozen=True)
+class UsedType:
+    parent_object_qualified_name: str
+    used_attribute_type: str
+
+
+@dataclasses.dataclass(frozen=True)
+class TypeTree:
+    token: str
+    children: List["TypeTree"]
+
+    @property
+    def is_fully_qualified_type(self) -> bool:
+        return len(self.children) == 0
+
+    def represent(self, level=0) -> str:
+        spacing = " " * 3 * level
+        if self.is_full_path_type:
+            return f"{spacing}{self.token}"
+        child_repr = "\n".join(child.represent(level=level + 1) for child in self.children)
+        return f"{spacing}{self.token} -> \n{child_repr}"
+
+    def __repr__(self) -> str:
+        return self.represent()
+
+
+def add_new_type_r(parent_tree: TypeTree, remaining_tokens: List[str]) -> None:
+    if remaining_tokens:
+        token = remaining_tokens.pop()
```

Comment:
suggestion: This mutates a shared object. If we want do this, let's add some warning in docstring.

Reply:
yeah, you are right 😄 was already on plan before merging, I've said that to @jjaakola-aiven before the review 😄 happy that also you noticed it, doing it now

Reply:
mmm I can transform it into an immutable version by "simply" replacing the variable with `token[:-2]` but IMO we aren't adding value and we are penalizing a lot the performance, consider that this object it's created and consumptioned by the method itself (another reason to make it private)

Reply:
That's fine, let's just add appropriate warning signs ;) 

<img src='https://images.theconversation.com/files/449630/original/file-20220302-17-a3llha.jpg?ixlib=rb-1.1.0&rect=0%2C366%2C3959%2C1976&q=45&auto=format&w=1356&h=668&fit=crop'>